java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/ArrowVectorIterator.java

```diff
@@ -0,0 +1,159 @@
+import java.util.Iterator;
+
+import org.apache.arrow.adapter.jdbc.consumer.CompositeJdbcConsumer;
+import org.apache.arrow.adapter.jdbc.consumer.JdbcConsumer;
+import org.apache.arrow.util.Preconditions;
+import org.apache.arrow.vector.VectorSchemaRoot;
+import org.apache.arrow.vector.types.pojo.Schema;
+
+/**
+ * VectorSchemaRoot iterator for partially converting JDBC data.
+ */
+public class ArrowVectorIterator implements Iterator<VectorSchemaRoot> {
+
+  private final ResultSet resultSet;
+  private final JdbcToArrowConfig config;
+
+  private final Schema schema;
+  private final ResultSetMetaData rsmd;
+
+  private final JdbcConsumer[] consumers;
+  private final CompositeJdbcConsumer compositeConsumer;
+  private boolean consumerCreated = false;
+
+  private VectorSchemaRoot nextBatch;
+
+  private final int targetBatchSize;
+
+  /**
+   * Construct an instance.
+   */
+  private ArrowVectorIterator(ResultSet resultSet, JdbcToArrowConfig config) throws SQLException {
+    this.resultSet = resultSet;
+    this.config = config;
+    this.schema = JdbcToArrowUtils.jdbcToArrowSchema(resultSet.getMetaData(), config);
+    this.targetBatchSize = config.getTargetBatchSize();
+
+    rsmd = resultSet.getMetaData();
+    consumers = new JdbcConsumer[rsmd.getColumnCount()];
+    this.compositeConsumer = new CompositeJdbcConsumer(consumers);
+  }
+
+  /**
+   * Create a ArrowVectorIterator to partially convert data.
+   */
+  public static ArrowVectorIterator create(
+      ResultSet resultSet,
+      JdbcToArrowConfig config)
+      throws SQLException {
+
+    ArrowVectorIterator iterator = new ArrowVectorIterator(resultSet, config);
+    try {
+      iterator.load();
+      return iterator;
+    } catch (Exception e) {
+      iterator.close();
+      throw new RuntimeException("Error occurs while creating iterator.", e);
+    }
+  }
+
+  // Loads the next schema root or null if no more rows are available.
+  private void load() throws IOException, SQLException {
+    VectorSchemaRoot root = null;
+    try {
+      root = VectorSchemaRoot.create(schema, config.getAllocator());
+      JdbcToArrowUtils.allocateVectors(root, JdbcToArrowUtils.DEFAULT_BUFFER_SIZE);
+    } catch (Exception e) {
+      if (root != null) {
+        root.close();
+      }
+      throw new RuntimeException("Error occurs while creating schema root.", e);
+    }
+
+    if (!consumerCreated) {
+      consumerCreated = true;
+      for (int i = 1; i <= consumers.length; i++) {
+        consumers[i - 1] = JdbcToArrowUtils.getConsumer(resultSet, i, resultSet.getMetaData().getColumnType(i),
+            root.getVector(rsmd.getColumnName(i)), config);
+      }
+    } else {
+      for (int i = 1; i <= consumers.length; i++) {
+        consumers[i - 1].resetValueVector(root.getVector(rsmd.getColumnName(i)));
+      }
+    }
+
+    // consume data
+    try {
+      int readRowCount = 0;
+      while (!((targetBatchSize != JdbcToArrowConfig.NO_LIMIT_BATCH_SIZE &&
```

Comment:
```suggestion
      while ((targetBatchSize == JdbcToArrowConfig.NO_LIMIT_BATCH_SIZE || readRowCount < targetBatchSize) && resultSet.next())
```
this reads more easily if you distribute the negation.